### ISSUE TITLE
protobuf: appliedPolicy not shown when Policy.NoAction is selected

### ISSUE THREAD

**Issue by 42wim** [PowerDNS/pdns#4384]
Working example with defpol=Policy.Custom

lua-config-file contains:

```
rpzMaster("xxxx", "dbl.rpz.spamhaus.org", {defpol=Policy.Custom, policyName="SpamHausDBL",localAddress="xxxxx",defcontent="www.example.be"})
protobufServer("127.0.0.1:2110", 2, 100, 2,32,128, false)
```

Protobuf dump

```
00000000  00 8b 08 02 12 10 77 34  30 2a 90 1d 4a b0 b1 a3  |......w40*..J...|
00000010  9f d1 5e 27 1f 96 20 01  28 01 32 04 7f 00 00 01  |..^'.. .(.2.....|
00000020  3a 04 7f 00 00 01 40 44  48 fd e8 96 be 05 50 fb  |:.....@DH.....P.|
00000030  b0 19 58 a1 d9 01 62 12  0a 0c 64 62 6c 74 65 73  |..X...b...dbltes|
00000040  74 2e 63 6f 6d 2e 10 01  18 01 6a 41 08 00 12 26  |t.com.....jA...&|
00000050  0a 0c 64 62 6c 74 65 73  74 2e 63 6f 6d 2e 10 05  |..dbltest.com...|
00000060  18 01 20 ac 02 2a 0f 77  77 77 2e 65 78 61 6d 70  |.. ..*.www.examp|
00000070  6c 65 2e 62 65 2e 1a 0b  53 70 61 6d 48 61 75 73  |le.be...SpamHaus|
00000080  44 42 4c 28 fd e8 96 be  05 30 ff af 19           |DBL(.....0...|
```

Failed with defpol=Policy.NoAction
lua-config-file contains:

```
rpzMaster("xxxx", "dbl.rpz.spamhaus.org", {defpol=Policy.NoAction, policyName="SpamHausDBL",localAddress="xxxxx",defcontent="www.example.be"})
protobufServer("127.0.0.1:2110", 2, 100, 2,32,128, false)
```

protobuf dump

```
00000000  00 73 08 02 12 10 58 a0  d8 ff 90 67 43 64 92 22  |.s....X....gCd."|
00000010  bb 63 63 9c f4 1a 20 01  28 01 32 04 7f 00 00 01  |.cc... .(.2.....|
00000020  3a 04 7f 00 00 01 40 38  48 c4 e8 96 be 05 50 eb  |:.....@8H.....P.|
00000030  b5 01 58 f0 bc 01 62 12  0a 0c 64 62 6c 74 65 73  |..X...b...dbltes|
00000040  74 2e 63 6f 6d 2e 10 01  18 01 6a 29 08 00 12 1b  |t.com.....j)....|
00000050  0a 0c 64 62 6c 74 65 73  74 2e 63 6f 6d 2e 10 01  |..dbltest.com...|
00000060  18 01 20 a0 52 2a 04 d9  46 b8 26 28 c4 e8 96 be  |.. .R*..F.&(....|
00000070  05 30 b0 b5 01                                    |.0...|
```

**Comment by rgacogne** [comment]
The result of a "query" policy that doesn't stop the processing (a NoAction or a Truncate over TCP) will be overridden by the result of the "post" policy and won't get logged. That could easily be fixed, but what do we do if the "post" policy matches too? Or even a NSDNAME/NSIP?

**Comment by 42wim** [comment]
Yes, you're right.

My use case is to see what impact enabling RPZ has on the users, by just logging the access and not applying the policy yet. I've got this working by logging the policyName from preresolve, but I was now trying to achieve the same with protobuf.

Maybe extending the protobuf message with something like "preAppliedPolicy" ? :)

**Comment by rgacogne** [comment]
You could add a policy tag based on the currently applied policy in `preresolve`, with something like:
```
function preresolve(dq)
  if dq.appliedPolicy.policyName then
    dq:addPolicyTag(dq.appliedPolicy.policyName)
  end
  return false
end
```

You will then get the policy tag in the protobuf message even if the applied policy is overridden later.

**Comment by 42wim** [comment]
Thanks for the tip, works perfect!

**Comment by rgacogne** [comment]
> The result of a "query" policy that doesn't stop the processing (a NoAction or a Truncate over TCP) will be overridden by the result of the "post" policy and won't get logged.

This was fixed in 4.3.0 by #8827.
